### PR TITLE
Missing bits from maistra-2.1

### PR DESCRIPTION
**Pull request by tedjpoole**

### CODE REVIEW

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: I wonder if emitting a warning in combination with a switch to `ENVOY_LOG_PERIODIC_TO_LOGGER` [1] would be better here. wdyt? 

[1] https://github.com/envoyproxy/envoy/blob/e727e1f50b33730a3a4fc061f95bcdd6d0c48768/source/common/common/logger.h#L541

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc:237]: So the format is that data is prefixed with a single byte, which represents the number of data bytes, correct? Can you add code-level comment to the method to document Filter::getAlpnProtocols while we are at it?

Tangentially related, but imho out of scope for now -- I'd love it if we could refactor this method into it's own unit of compilation under `/maistra/source/tls/alpn_util.cc` or some such.. and then we could have unit tests for this in `/maistra/source/tls/test_alpn_util.cc`.  That would prove correctness, tease out issues when run under asan, as well as serve as semi-documentation on assumptions and edge cases etc.

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: should we put a cap on the value of len that we are willing to accept?

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: ditto

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc:250]: readability nit: afaict this increment could move into the usual position in the for() loop on the premise that we declare it outside of that and initialize it to zero -- but that probably is just personal preference

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (oschaaf) [inline source/extensions/transport_sockets/tls/context_impl.cc]: Did we verify upstream also stripped this? Or is this specific to us?

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (oschaaf) [inline source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:168]: Are we sure we included the rest of Marko his fix for this validation?

**Review by oschaaf — CHANGES_REQUESTED** [review]
Nice work, like how readable `getAlpnProtocols` got. Left a few nits/questions/comments.

File: bazel/foreign_cc/luajit2.patch
Comment (jwendell) [inline bazel/foreign_cc/luajit2.patch]: We removed this in a recent pr

File: bazel/repositories.bzl
Comment (jwendell) [inline bazel/repositories.bzl:1009]: Are you sure? I think this patch is already present on 2.2

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: I wasn't too sure about logging level to use. I had a look around and it seemed to me like warning/error log messages are generally used to report on the (mal)functioning of envoy itself, rather than to report on erroneous content flowing through envoy, but I'll change it as suggested.

File: bazel/repositories.bzl
Comment (knm3000) [reply bazel/repositories.bzl:1009]: It is currently not present in 2.2: some luajit2 changes have not been merged from 2.1 into 2.2. This PR contains all missing luajit2 code that is needed to build envoy/proxy with luajit2.

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: At the point where getAlpnProtocols() is called the buffer has already been allocated and populated off the wire, so why would we reject what's in it?

Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: makes sense resolving

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (tedjpoole) [reply source/extensions/transport_sockets/tls/context_impl.cc]: This is specific to us and the decision to remove it was made in discussion with @twghu

File: source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc
Comment (tedjpoole) [reply source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc:168]: Yes, it was only the change to the comment that didn't make it to 2.2

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (oschaaf) [reply source/extensions/transport_sockets/tls/context_impl.cc]: great, thanks for sharing context. perhaps leave note for our future selves in the code? e.g. something like:

```
// MAISTRA - removed "if (response->isExpired()" handling, because .. 
```

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: The thing is that with regular logging this could result in far too much log spew. But i think that hitting these edge cases is worth surfacing to the user. With the periodic log function, we don't be flooding the logs while still surfacing that this stuff is happening. That was my reasoning.

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: shall we init protocol_length to 0? :-)

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [inline source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: perfect, thank you for adding this. 
just one nit: s/underun/underrun/

**Review by oschaaf — CHANGES_REQUESTED** [review]

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: It didn't look necessary to me because it does get initialised before use.

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: Done

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: Sorry if I'm a little dense, but where does protocol_length get initialised? 
(From what I can tell `i += protocol_length` is the first statement that mentions it?)

Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: The `i += protocol_length` statement doesn't get executed until after the first run through the loop, so the `protocol_length = data[i++]` statement is the first to be executed.

Comment (jwendell) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: evil :D

Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: Aaah yes, I missed that; thanks for explaining. Resolving this..

Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: though thinking about it, perhaps a one line comment about this won't hurt here just to make it obvious to future readers?

Comment (jwendell) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: or just initialize it... compiler might optimize it anyway, so...

We are also doing other evil stuff like changing the for loop variable in the middle of the loop. It might work, but my personal opinion is that in such cases we could just use while instead of for.

Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: So I mulled that over a bit, and now I agree on that I'd prefer a while loop as well. It seems like we're going through hoops to try and fit what we are doing into the original for loop construction, which I think started us off on the wrong foot: a while loop seems to be a more natural intuitive candidate here. @tedjpoole sorry about the flipflopping :-) curious to hear what you think?

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: Done

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: done

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (tedjpoole) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc:237]: Added comment to explain the format of the data being unmarshalled.

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (tedjpoole) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Reinstated this check because it looks like it has some value. What do you think @twghu? Should I also add it upstream?

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc:237]: 👍 thank you, resolving.

File: bazel/foreign_cc/luajit2.patch
Comment (jwendell) [reply bazel/foreign_cc/luajit2.patch]: This change needs to be removed otherwise it will fail and we will only notice it on ppc builds, not on prow (x86). This is an issue: We currently don't have a way to try builds on non-x86 platforms before the pr is merged.

File: source/extensions/filters/listener/tls_inspector/tls_inspector.cc
Comment (oschaaf) [reply source/extensions/filters/listener/tls_inspector/tls_inspector.cc]: thanks for following up, I like the while loop better. Resolving
/cc @jwendell

**Review by oschaaf — APPROVED** [review]
modulo pending comments from others, this LGTM

File: bazel/repositories.bzl
Comment (jwendell) [reply bazel/repositories.bzl:1009]: True. I just hit this in a ppc product build. The build fails because of this.

Comment (tedjpoole) [reply bazel/repositories.bzl:1009]: @jwendell just to be clear, do you mean that you want this left in?

Comment (jwendell) [reply bazel/repositories.bzl:1009]: Yes. This is missing in 2.2 indeed.

You just need to get rid of the change to the flags mentioned above
```diff
-CCOPT= -O2 -fomit-frame-pointer
+CCOPT= -O2 -fomit-frame-pointer -fPIC -fPIE -fplugin=annobin -fstack-protector-strong -fstack-protector-all
```

File: bazel/foreign_cc/luajit2.patch
Comment (tedjpoole) [reply bazel/foreign_cc/luajit2.patch]: Done. It builds & tests ok (on x86) using the `--//source/extensions/filters/common/lua:luajit2=1` option.

Comment (jwendell) [reply bazel/foreign_cc/luajit2.patch]: Awesome, thanks.

**Review by jwendell — APPROVED** [review]

File: source/extensions/transport_sockets/tls/context_impl.cc
Comment (twghu) [reply source/extensions/transport_sockets/tls/context_impl.cc]: The expiry check if() was added 
on 13/4/21 https://github.com/maistra/envoy/commit/412c091b302cc688bf3269022bf679bced828b2f#diff-2b39f212c2f760ae547351a46f92ad08c0ec84c556d678ebe9388b9d008e7313

and was removed on 20/5/21
https://github.com/maistra/envoy/commit/963afc8094dc2c17bd24526f2de11841a5904b1b

The last commit should indicate reason for removal, a sanity check would be a good idea.

Comment (oschaaf) [reply source/extensions/transport_sockets/tls/context_impl.cc]: So digging into this, it seems the check is (also) done in [ServerContextImpl::ocspStapleAction](https://github.com/maistra/envoy/blob/0428980c9210b2c7de98eb64e73fcfc526c96426/source/extensions/transport_sockets/tls/context_impl.cc#L948).  So I think we can remove it.

**Pull request merged**
